datafusion-cli/src/object_storage.rs

```diff
@@ -563,6 +563,592 @@ pub(crate) async fn get_object_store(
+            }
+        }
+    }
+
+    impl fmt::Display for RequestSummary {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            writeln!(f, "  count: {}", self.count)?;
+
+            if let Some(dur_stats) = &self.duration_stats {
+                writeln!(f, "  duration min: {:.6}s", dur_stats.min.as_secs_f32())?;
+                writeln!(f, "  duration max: {:.6}s", dur_stats.max.as_secs_f32())?;
+                let avg = dur_stats.sum.as_secs_f32() / (self.count as f32);
+                writeln!(f, "  duration avg: {:.6}s", avg)?;
+            }
+
+            if let Some(size_stats) = &self.size_stats {
+                writeln!(f, "  size min: {} B", size_stats.min)?;
+                writeln!(f, "  size max: {} B", size_stats.max)?;
+                let avg = size_stats.sum / self.count;
+                writeln!(f, "  size avg: {} B", avg)?;
+                writeln!(f, "  size sum: {} B", size_stats.sum)?;
+            }
+
+            Ok(())
+        }
+    }
+
+    struct Stats<T: Copy + Ord + AddAssign<T>> {
+        min: T,
+        max: T,
+        sum: T,
+    }
+
+    impl<T: Copy + Ord + AddAssign<T>> Stats<T> {
+        fn push(&mut self, val: T) {
+            self.min = cmp::min(val, self.min);
+            self.max = cmp::max(val, self.max);
+            self.sum += val;
+        }
+    }
+
+    impl default::Default for Stats<Duration> {
+        fn default() -> Self {
+            Self {
+                min: Duration::MAX,
+                max: Duration::ZERO,
+                sum: Duration::ZERO,
+            }
+        }
+    }
+
+    impl default::Default for Stats<usize> {
+        fn default() -> Self {
+            Self {
+                min: usize::MAX,
+                max: usize::MIN,
+                sum: 0,
+            }
+        }
+    }
+
+    #[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+    pub enum InstrumentedObjectStoreMode {
+        #[default]
+        Disabled,
+        Summary,
+        Trace,
+    }
+
+    impl fmt::Display for InstrumentedObjectStoreMode {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "{:?}", self)
+        }
+    }
+
+    impl FromStr for InstrumentedObjectStoreMode {
+        type Err = DataFusionError;
+
+        fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+            match s.to_lowercase().as_str() {
+                "disabled" => Ok(Self::Disabled),
+                "summary" => Ok(Self::Summary),
+                "trace" => Ok(Self::Trace),
+                _ => Err(DataFusionError::Execution(format!("Unrecognized mode {s}"))),
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct InstrumentedObjectStore {
+        inner: Arc<dyn ObjectStore>,
+        instrument_mode: AtomicU8,
+        requests: Mutex<Vec<RequestDetails>>,
+    }
+
+    impl fmt::Display for InstrumentedObjectStore {
```

Comment:
I found it a bit confusing that `Display` also modified the object store -- maybe we could have an explicit `take/display` function 🤔